File: src/main/java/com/github/trcdeveloppers/clayium/blocks/machines/clayworktable/ClayWorkTableRecipes.java
package com.github.trcdeveloppers.clayium.blocks.machines.clayworktable;

import com.github.trcdeveloppers.clayium.items.ClayiumItems;
import net.minecraft.init.Items;
import net.minecraft.item.ItemStack;
import net.minecraft.util.NonNullList;

import java.util.List;
import java.util.stream.Collectors;

public class ClayWorkTableRecipes {
    static final NonNullList<Recipe> recipes = NonNullList.create();

    public static void init() {
        addRecipe(new ItemStack(Items.CLAY_BALL), new ItemStack(ClayiumItems.getItem("clay_stick")),
            ClayWorkTableMethod.ROLLING_HAND, 4);
        addRecipe(new ItemStack(ClayiumItems.getItem("clay_plate"), 3), new ItemStack(ClayiumItems.getItem("large_clay_ball")),
            ClayWorkTableMethod.ROLLING_HAND, 40);
        addRecipe(new ItemStack(ClayiumItems.getItem("large_clay_ball")), new ItemStack(ClayiumItems.getItem("clay_cylinder")),
            ClayWorkTableMethod.ROLLING_HAND, 4);
        addRecipe(new ItemStack(ClayiumItems.getItem("clay_cylinder")), new ItemStack(ClayiumItems.getItem("clay_needle")),
            ClayWorkTableMethod.ROLLING_HAND, 3);

        addRecipe(new ItemStack(ClayiumItems.getItem("clay_plate")), new ItemStack(ClayiumItems.getItem("clay_blade")),
            ClayWorkTableMethod.PUNCH, 10);
        //todo: raw_clay_slicerの実装
//        addRecipe(new ItemStack(ClayiumItems.getItem("clay_disc")), new ItemStack(ClayiumItems.getItem("raw_clay_slicer")),
//            ClayWorkTableMethod.PUNCH, 15);
        addRecipe(new ItemStack(ClayiumItems.getItem("large_clay_ball")), new ItemStack(ClayiumItems.getItem("clay_disc")),
            ClayWorkTableMethod.PUNCH, 30);

        addRecipe(new ItemStack(ClayiumItems.getItem("clay_plate"), 6), new ItemStack(ClayiumItems.getItem("large_clay_plate")),
            ClayWorkTableMethod.ROLLING_PIN, 10);
        addRecipe(new ItemStack(ClayiumItems.getItem("clay_plate"), 6), new ItemStack(ClayiumItems.getItem("clay_blade")),
            new ItemStack(Items.CLAY_BALL, 2), ClayWorkTableMethod.ROLLING_PIN, 10);
//        addRecipe(new ItemStack(ClayiumItems.getItem("clay_disc")), new ItemStack(ClayiumItems.getItem("raw_clay_slicer")),
//            ClayWorkTableMethod.ROLLING_PIN, 2);
        addRecipe(new ItemStack(ClayiumItems.getItem("large_clay_ball"), 6), new ItemStack(ClayiumItems.getItem("clay_disc")),
            new ItemStack(Items.CLAY_BALL, 2), ClayWorkTableMethod.ROLLING_PIN, 4);

        addRecipe(new ItemStack(ClayiumItems.getItem("clay_disc"), 6), new ItemStack(ClayiumItems.getItem("clay_plate")),
            new ItemStack(Items.CLAY_BALL, 2), ClayWorkTableMethod.CUT_PLATE, 4);

        addRecipe(new ItemStack(ClayiumItems.getItem("clay_disc"), 6), new ItemStack(ClayiumItems.getItem("clay_ring")),
            new ItemStack(ClayiumItems.getItem("clay_small_disc")), ClayWorkTableMethod.CUT_DISC, 2);
        addRecipe(new ItemStack(ClayiumItems.getItem("clay_small_disc"), 6), new ItemStack(ClayiumItems.getItem("clay_small_ring")),
            new ItemStack(ClayiumItems.getItem("clay_short_stick")), ClayWorkTableMethod.CUT_DISC, 1);

        addRecipe(new ItemStack(ClayiumItems.getItem("clay_plate")), new ItemStack(ClayiumItems.getItem("clay_stick")),
            ClayWorkTableMethod.CUT, 3);
        addRecipe(new ItemStack(ClayiumItems.getItem("clay_cylinder")), new ItemStack(ClayiumItems.getItem("clay_small_disc"), 8),
            ClayWorkTableMethod.CUT, 7);
    }

    //todo: inputとmethodが両方同じときの対処
    private static void addRecipe(ItemStack input, ItemStack primaryOutput, ItemStack secondaryOutput, ClayWorkTableMethod method, int clicks) {
        if (!getRecipeFor(input, method).isEmpty()) {
            //todo: add logger, and warn
            System.out.println("duplicate recipe");
            return;
        }
        recipes.add(new Recipe(input, primaryOutput, secondaryOutput, method, clicks));
    }
    private static void addRecipe(ItemStack input, ItemStack primaryOutput, ClayWorkTableMethod method, int clicks) {
        addRecipe(input, primaryOutput, ItemStack.EMPTY, method, clicks);
    }

    static List<Recipe> getRecipesFor(ItemStack input) {
        return recipes.stream()
            .filter(recipe -> recipe.INPUT.isItemEqual(input))
            .filter(recipe -> recipe.INPUT.getCount() <= input.getCount())
            .collect(Collectors.toList());
    }

    static Recipe getRecipeFor(ItemStack input, ClayWorkTableMethod method) {
        for (Recipe recipe : recipes) {
            if (!(recipe.INPUT.isItemEqual(input) && (recipe.INPUT.getCount() <= input.getCount()) && (recipe.METHOD == method))) {
                continue;
            }
            return recipe;
        }
        return Recipe.EMPTY;
    }

    static class Recipe {

        public static final Recipe EMPTY = new Recipe(ItemStack.EMPTY, ItemStack.EMPTY, ItemStack.EMPTY, ClayWorkTableMethod.EMPTY, 0);

        public final ItemStack INPUT;
        public final ItemStack OUTPUT_1;
        public final ItemStack OUTPUT_2;
        public final ClayWorkTableMethod METHOD;
        public final int CLICKS;
        public Recipe(ItemStack input, ItemStack primaryOutput, ItemStack secondaryOutput, ClayWorkTableMethod method, int clicks) {
            this.INPUT = input;
            this.OUTPUT_1 = primaryOutput;
            this.OUTPUT_2 = secondaryOutput;
            this.METHOD = method;
            this.CLICKS = clicks;
        }

        public boolean hasSecondaryOutput() {
            return !this.OUTPUT_2.isEmpty();
        }

        public boolean isEmpty() {
            return this == Recipe.EMPTY;
        }

        @Override
        public String toString() {
            return "Recipe{" +
                    "INPUT=" + INPUT +
                    ", OUTPUT_1=" + OUTPUT_1 +
                    ", OUTPUT_2=" + OUTPUT_2 +
                    ", METHOD=" + METHOD +
                    ", CLICKS=" + CLICKS +
                    '}';
        }
    }
}

File: src/main/kotlin/com/github/trcdevelopers/clayium/common/recipe/loader/ClayWorkTableRecipes.kt
package com.github.trcdevelopers.clayium.common.recipe.loader

import com.github.trcdevelopers.clayium.common.blocks.clayworktable.ClayWorkTableMethod.CUT
import com.github.trcdevelopers.clayium.common.blocks.clayworktable.ClayWorkTableMethod.CUT_DISC
import com.github.trcdevelopers.clayium.common.blocks.clayworktable.ClayWorkTableMethod.CUT_PLATE
import com.github.trcdevelopers.clayium.common.blocks.clayworktable.ClayWorkTableMethod.PUNCH
import com.github.trcdevelopers.clayium.common.blocks.clayworktable.ClayWorkTableMethod.ROLLING_HAND
import com.github.trcdevelopers.clayium.common.blocks.clayworktable.ClayWorkTableMethod.ROLLING_PIN
import com.github.trcdevelopers.clayium.common.items.metaitem.MetaItemClayParts.CLAY_BLADE
import com.github.trcdevelopers.clayium.common.items.metaitem.MetaItemClayParts.CLAY_CYLINDER
import com.github.trcdevelopers.clayium.common.items.metaitem.MetaItemClayParts.CLAY_DISC
import com.github.trcdevelopers.clayium.common.items.metaitem.MetaItemClayParts.CLAY_NEEDLE
import com.github.trcdevelopers.clayium.common.items.metaitem.MetaItemClayParts.CLAY_RING
import com.github.trcdevelopers.clayium.common.items.metaitem.MetaItemClayParts.CLAY_SHORT_STICK
import com.github.trcdevelopers.clayium.common.items.metaitem.MetaItemClayParts.CLAY_SMALL_DISC
import com.github.trcdevelopers.clayium.common.items.metaitem.MetaItemClayParts.CLAY_SMALL_RING
import com.github.trcdevelopers.clayium.common.items.metaitem.MetaItemClayParts.CLAY_STICK
import com.github.trcdevelopers.clayium.common.items.metaitem.MetaItemClayParts.LARGE_CLAY_BALL
import com.github.trcdevelopers.clayium.common.recipe.CWTRecipes
import com.github.trcdevelopers.clayium.common.unification.OreDictUnifier
import com.github.trcdevelopers.clayium.common.unification.material.CMaterials
import com.github.trcdevelopers.clayium.common.unification.ore.OrePrefix
import net.minecraft.init.Items
import net.minecraft.item.ItemStack


object ClayWorkTableRecipes {
    fun registerRecipes() {
        CWTRecipes.CLAY_WORK_TABLE.register {
            input(CLAY_CYLINDER)
            outputs(CLAY_NEEDLE.getStackForm(1))
            method(ROLLING_HAND)
            clicks(3)
        }
        CWTRecipes.CLAY_WORK_TABLE.register {
            input(Items.CLAY_BALL)
            outputs(CLAY_STICK.getStackForm(1))
            method(ROLLING_HAND)
            clicks(4)
        }
        CWTRecipes.CLAY_WORK_TABLE.register {
            input(LARGE_CLAY_BALL)
            outputs(CLAY_CYLINDER.getStackForm(1))
            method(ROLLING_HAND)
            clicks(5)
        }
        CWTRecipes.CLAY_WORK_TABLE.register {
            input(OrePrefix.plate, CMaterials.clay, 3)
            outputs(LARGE_CLAY_BALL.getStackForm(1))
            method(ROLLING_HAND)
            clicks(40)
        }

        CWTRecipes.CLAY_WORK_TABLE.register {
            input(OrePrefix.largePlate, CMaterials.clay)
            outputs(CLAY_BLADE.getStackForm(1))
            method(PUNCH)
            clicks(10)
        }
//        CWTRecipes.CLAY_WORK_TABLE.register {
//            input(CLAY_DISC)
//            outputs(/* raw clay slicer */)
//            method(PUNCH)
//            clicks(15)
//        }
        CWTRecipes.CLAY_WORK_TABLE.register {
            input(LARGE_CLAY_BALL)
            outputs(CLAY_DISC.getStackForm(1))
            method(PUNCH)
            clicks(30)
        }

        CWTRecipes.CLAY_WORK_TABLE.register {
            input(OrePrefix.plate, CMaterials.clay)
            outputs(CLAY_BLADE.getStackForm(1), ItemStack(Items.CLAY_BALL, 2))
            method(ROLLING_PIN)
            clicks(1)
        }
        CWTRecipes.CLAY_WORK_TABLE.register {
            input(LARGE_CLAY_BALL)
            outputs(CLAY_DISC.getStackForm(1), ItemStack(Items.CLAY_BALL, 2))
            method(ROLLING_PIN)
            clicks(4)
        }
        CWTRecipes.CLAY_WORK_TABLE.register {
            input(OrePrefix.plate, CMaterials.clay, 6)
            outputs(OreDictUnifier.get(OrePrefix.largePlate, CMaterials.clay))
            method(ROLLING_PIN)
            clicks(10)
        }
//        CWTRecipes.CLAY_WORK_TABLE.register {
//            input(CLAY_DISC)
//            outputs(/* raw clay slicer */)
//            method(ROLLING_PIN)
//            clicks(2)
//        }

        CWTRecipes.CLAY_WORK_TABLE.register {
            input(CLAY_DISC)
            outputs(OreDictUnifier.get(OrePrefix.plate, CMaterials.clay), ItemStack(Items.CLAY_BALL, 2))
            method(CUT_PLATE)
            clicks(4)
        }

        CWTRecipes.CLAY_WORK_TABLE.register {
            input(CLAY_SMALL_DISC)
            outputs(CLAY_SMALL_RING.getStackForm(1), CLAY_SHORT_STICK.getStackForm(1))
            method(CUT_DISC)
            clicks(1)
        }
        CWTRecipes.CLAY_WORK_TABLE.register {
            input(CLAY_DISC)
            outputs(CLAY_RING.getStackForm(1), CLAY_SMALL_DISC.getStackForm(1))
            method(CUT_DISC)
            clicks(2)
        }

        CWTRecipes.CLAY_WORK_TABLE.register {
            input(OrePrefix.plate, CMaterials.clay)
            outputs(CLAY_STICK.getStackForm(4))
            method(CUT)
            clicks(3)
        }
        CWTRecipes.CLAY_WORK_TABLE.register {
            input(CLAY_CYLINDER)
            outputs(CLAY_SMALL_DISC.getStackForm(8))
            method(CUT)
            clicks(7)
        }
    }
}